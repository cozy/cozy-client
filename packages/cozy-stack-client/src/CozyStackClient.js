import AppToken from './AppToken'
import DocumentCollection from './DocumentCollection'
import FileCollection from './FileCollection'

const normalizeUri = uri => {
  while (uri[uri.length - 1] === '/') {
    uri = uri.slice(0, -1)
  }
  return uri
}

/**
 * Cozy stack HTTP client.
 */
export default class CozyStackClient {
  constructor({ token, uri = '' }) {
    this.uri = normalizeUri(uri)
    this.setCredentials(token)
  }

  /**
   * Creates a {@link DocumentCollection} instance.
   *
   * @param  {String} doctype The collection doctype.
   * @return {DocumentCollection}
   */
  collection(doctype) {
    if (!doctype) {
      throw new Error('CozyStackClient.collection() called without a doctype')
    }
    switch (doctype) {
      case 'io.cozy.files':
        return new FileCollection(doctype, this)
      default:
        return new DocumentCollection(doctype, this)
    }
  }

  /**
   * Fetches JSON in an authorized way.
   *
   * @param  {String} method The HTTP method.
   * @param  {String} path The URI.
   * @param  {Object} body The payload.
   * @param  {Object} options
   * @return {Object}
   * @throws {FetchError}
   */
  async fetch(method, path, body, options = {}) {
    options.method = method
    const headers = (options.headers = options.headers || {})

    headers['Accept'] = 'application/json'

    if (method !== 'GET' && method !== 'HEAD' && body !== undefined) {
      if (headers['Content-Type']) {
        options.body = body
      } else {
        headers['Content-Type'] = 'application/json'
        options.body = JSON.stringify(body)
      }
    }

    const credentials = options.credentials || this.getCredentials()
    if (credentials) {
      headers['Authorization'] = credentials
      // the option credentials:include tells fetch to include the cookies in the
      // request even for cross-origin requests
      options.credentials = 'include'
    }

    const resp = await fetch(this.fullpath(path), options)
    const contentType = resp.headers.get('content-type')
    const isJson = contentType && contentType.indexOf('json') >= 0
    const data = await (isJson ? resp.json() : resp.text())
    if (resp.ok) {
      return data
    }
    throw new FetchError(resp, data)
  }

  fullpath(path) {
    return this.uri + path
  }

  getCredentials() {
    return this.token ? this.token.toAuthHeader() : null
  }

  setCredentials(token) {
    this.token = token ? new AppToken(token) : null
  }
}

export class FetchError extends Error {
  constructor(response, reason) {
    super()
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    }
    // WARN We have to hardcode this because babel doesn't play nice when extending Error
    this.name = 'FetchError'
    this.response = response
    this.url = response.url
    this.status = response.status
    this.reason = reason

    Object.defineProperty(this, 'message', {
      value:
        reason.message ||
        (typeof reason === 'string' ? reason : JSON.stringify(reason))
    })
  }
}
