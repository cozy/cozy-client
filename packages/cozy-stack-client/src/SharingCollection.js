import DocumentCollection, { normalizeDoc } from './DocumentCollection'
import { isFile, isDirectory } from './FileCollection'
import { uri } from './utils'

const normalizeSharing = sharing => normalizeDoc(sharing, 'io.cozy.sharings')

export default class SharingCollection extends DocumentCollection {
  async findByDoctype(doctype) {
    const resp = await this.client.fetch(
      'GET',
      uri`/sharings/doctype/${doctype}`
    )
    return {
      ...resp,
      data: resp.data.map(normalizeSharing)
    }
  }

  async share(document, recipients, sharingType, description) {
    const resp = await this.client.fetch('POST', '/sharings/', {
      data: {
        type: 'io.cozy.sharings',
        attributes: {
          description,
          open_sharing: sharingType === 'two-way',
          rules: getSharingRules(document, sharingType)
        },
        relationships: {
          recipients: {
            data: recipients.map(({ _id, _type }) => ({ id: _id, type: _type }))
          }
        }
      }
    })
    return { data: normalizeSharing(resp.data) }
  }

  async addRecipients(sharing, recipients, sharingType) {
    const recipientsPayload = {
      data: recipients.map(({ _id, _type }) => ({
        id: _id,
        type: _type
      }))
    }
    const resp = await this.client.fetch(
      'POST',
      uri`/sharings/${sharing._id}/recipients`,
      {
        data: {
          type: 'io.cozy.sharings',
          id: sharing._id,
          relationships:
            sharingType === 'two-way'
              ? {
                  recipients: recipientsPayload
                }
              : {
                  read_only_recipients: recipientsPayload
                }
        }
      }
    )
    return { data: normalizeSharing(resp.data) }
  }

  revokeRecipient(sharing, recipientEmail) {
    const memberIndex = sharing.attributes.members.findIndex(
      m => m.email === recipientEmail
    )
    return this.client.fetch(
      'DELETE',
      uri`/sharings/${sharing._id}/recipients/${memberIndex}`
    )
  }

  revokeSelf(sharing) {
    return this.client.fetch(
      'DELETE',
      uri`/sharings/${sharing._id}/recipients/self`
    )
  }
}

// Rules determine the behavior of the sharing when changes are made to the shared document
// See https://github.com/cozy/cozy-stack/blob/master/docs/sharing-design.md#description-of-a-sharing
const getSharingRules = (document, sharingType) => {
  const { _id, _type } = document
  return isFile(document)
    ? [
        {
          title: document.name,
          doctype: 'io.cozy.files',
          values: [_id],
          ...getSharingPolicy(document, sharingType)
        }
      ]
    : [
        {
          title: 'collection',
          doctype: _type,
          values: [_id],
          ...getSharingPolicy(document, sharingType)
        },
        {
          title: 'items',
          doctype: 'io.cozy.files',
          values: [`${_type}/${_id}`],
          selector: 'referenced_by',
          ...(sharingType === 'two-way'
            ? { add: 'sync', update: 'sync', remove: 'sync' }
            : { add: 'push', update: 'none', remove: 'push' })
        }
      ]
}

const getSharingPolicy = (document, sharingType) => {
  if (isFile(document) && isDirectory(document)) {
    return sharingType === 'two-way'
      ? { add: 'sync', update: 'sync', remove: 'sync' }
      : { add: 'push', update: 'push', remove: 'push' }
  }
  return sharingType === 'two-way'
    ? { update: 'sync', remove: 'revoke' }
    : { update: 'push', remove: 'revoke' }
}
