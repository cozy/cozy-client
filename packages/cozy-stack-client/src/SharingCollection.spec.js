jest.mock('./CozyStackClient')

import CozyStackClient from './CozyStackClient'
import SharingCollection from './SharingCollection'

const FOLDER = {
  _type: 'io.cozy.files',
  _id: 'folder_1',
  type: 'directory',
  name: 'bills'
}

const RECIPIENT = {
  _type: 'io.cozy.contacts',
  _id: 'contact_1',
  attributes: {
    email: [{ address: 'jdoe@doe.com', primary: true }]
  }
}

describe('SharingCollection', () => {
  const client = new CozyStackClient()
  const collection = new SharingCollection('io.cozy.sharings', client)

  describe('findByDoctype', () => {
    beforeAll(() => {
      client.fetchJSON.mockReturnValue(Promise.resolve({ data: [] }))
    })

    it('should call the right route', async () => {
      await collection.findByDoctype('io.cozy.files')
      expect(client.fetchJSON).toHaveBeenCalledWith(
        'GET',
        '/sharings/doctype/io.cozy.files'
      )
    })
  })

  describe('share', () => {
    beforeAll(() => {
      client.fetch.mockReset()
      client.fetchJSON.mockReturnValue(Promise.resolve({ data: [] }))
    })

    it('should call the right route with the right payload', async () => {
      await collection.share(FOLDER, [RECIPIENT], 'one-way', 'foo')
      expect(client.fetchJSON).toMatchSnapshot()
    })

    it('should call the right route with the right payload', async () => {
      await collection.share(FOLDER, [RECIPIENT], 'one-way', 'foo', '/preview')
      expect(client.fetchJSON).toMatchSnapshot()
    })
  })
})
