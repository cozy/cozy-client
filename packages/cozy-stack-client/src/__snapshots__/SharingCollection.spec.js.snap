// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharingCollection share should call the right route with the right payload 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "GET",
      "/sharings/doctype/io.cozy.files",
    ],
    Array [
      "POST",
      "/sharings/",
      Object {
        "data": Object {
          "attributes": Object {
            "description": "foo",
            "open_sharing": false,
            "rules": Array [
              Object {
                "add": "push",
                "doctype": "io.cozy.files",
                "remove": "push",
                "title": "bills",
                "update": "push",
                "values": Array [
                  "folder_1",
                ],
              },
            ],
          },
          "relationships": Object {
            "recipients": Object {
              "data": Array [
                Object {
                  "id": "contact_1",
                  "type": "io.cozy.contacts",
                },
              ],
            },
          },
          "type": "io.cozy.sharings",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
